Add vSCMI support to the PM branch

lorc opened the pull request.

lorc committed:
vscpi: add vscmi server implementation

Signed-off-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>

lorc committed:
rcar: create rcar3 platform file

So it can call SMPI-triggered mailbox

Signed-off-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>

lorc committed:
vscmi: initialize vcpu OPPs on domain startup

Signed-off-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>

lorc committed:
vscmi: create FDT nodes for vscmi needs

Signed-off-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>

Commit:
xen/device-tree: Add ability to handle nodes with interrupts-extended prop

The "interrupts-extended" property is a special form for use when
a node needs to reference multiple interrupt parents.

According to the:
Linux/Documentation/devicetree/bindings/interrupt-controller/interrupts.txt

But, there are cases when "interrupts-extended" property is used for
"outside /soc node" with a single interrupt parent as an equivalent of
pairs ("interrupt-parent" + "interrupts").

The good example here is ARCH timer node for R-Car Gen3/Gen2 family,
which is mandatory device for Xen usage on ARM. And without ability
to handle such nodes, Xen fails to operate.

So, this patch adds required support for Xen to be able to handle
nodes with that property.

Signed-off-by: Oleksandr Tyshchenko <oleksandr_tyshchenko@epam.com>

Reviewed-by: Stefano Stabellini <sstabellini@kernel.org>

lorc committed:
xen: vscmi: move vscmi memory initialization to vscmi.c

This is a preparation for domctl calls support.
Allocate and map shared memory into domain p2m in dedicated function
domain_vscmi_init().

Unmap and free the shared buffer in domain_vscmi_free().

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
xen: add domctl interface for enabling vscmi

This call is used to enable vscmi for a given domain, so
not only Dom0 can use SCMI.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
tools: domctl: add xc_dom_vscmi_init() function

This is toolstack-side function for enabling vscmi for
a new domains.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
tools: add vscmi= option to xl.cfg

Now, you can add vscmi=1 to domain config file to enable
vscmi functionality.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
tools: create SCMI nodes in the DT if scmi is enabled

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
vscmi: rework opp table mode

Now there are 16 virtual OPPs and custom function to scale
platform's frequencies.
This provides much more fine control for guests and partially
eliminates problem with non-uniform scale of real platform
OPPs.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
vscmi: add notifier chain

So we can notify other parts of XEN about guest's requests

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
cpufreq_vscmi: add initial implementation of the governor

This is very basic implementation of vscmi governor. It listens
to requests from guests and sets the frequency to the last
requested one.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
vscmi: add vscmi_enabled_for_domain() helper function

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
vscmi: fix vscmi_scale_opp rounding errors

If we don't want to get rounding errors it is better
to multiply first and divide then. This will greatly
help on diapason ends.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
vscmi: start perf level from 1, not from 0

Looks like Linux does not performance level 0 and never
uses it. So wee need to start counting from 1.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
vscmi: tell guest that we are using least possible opp by default

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
cpufreq_vscmi: set freq for all pcpus based on max requested opp

This patch implements base governing policy for cpufreq_vscmi
governor: we are choosing biggest requested OPP as a base freq.

We are setting it to all pCPUs that requested vCPU can run.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

otyshchenko1 commented on xen/include/public/device_tree_defs.h:
```diff
@@ -8,6 +8,10 @@
+#define GUEST_PHANDLE_SMCPI_MBOX        64000
+#define GUEST_PHANDLE_SMCPI_SHMEM       64001
+#define GUEST_PHANDLE_SMCPI_CPU_CLOCKS  64002
+
```
s/SMCPI/SCMI for everything?

lorc replied:
Yeah, good catch. Thank you.

otyshchenko1 commented on xen/arch/arm/vscmi.c:
```diff
@@ -0,0 +1,356 @@
+    spin_unlock(&add_remove_lock);
+}
+
+static void write_32(u8 *ptr, int32_t data)
```
Do we have something already existing in Xen?

otyshchenko1 commented on xen/arch/arm/vscmi.c:
```diff
@@ -0,0 +1,356 @@
+#define PERF_OPP_COUNT    16            /* Sadly, Linux does not support more */
+#define PERF_OPPS_PER_CALL 5
+
+static DEFINE_SPINLOCK(add_remove_lock);
```
Please consider adding a commit describing such a big lock

lorc replied:
Why "big"? It is need to protect one variable - `vscmi_chain`

otyshchenko1 commented on xen/arch/arm/vscmi.c:
```diff
@@ -0,0 +1,356 @@
+
+#define SCMI_VERSION 0x10000
+/* One agent, one protocol */
+#define SCMI_PROTO_ATTRS ((1 << 0) | (1 << 8))
```
Can't we use BIT marcos here?

otyshchenko1 commented on tools/libxl/libxl_arm.c:
```diff
@@ -293,6 +293,133 @@ static int make_root_properties(libxl__gc *gc,
+{
+    int res;
+    char name[32];
+    /* TODO: This is the ugly hack, because dom->vscmi_gfn is not populated right now */
```
Who and when will address this?

lorc replied:
We. In some time in the future.  This is why I added this comment with `TODO`

otyshchenko1 commented on xen/drivers/cpufreq/Makefile:
```diff
@@ -1,4 +1,5 @@
 obj-y += cpufreq.o
 obj-y += cpufreq_ondemand.o
 obj-y += cpufreq_misc_governors.o
+obj-y += cpufreq_vscmi.o
```
General question regarding VSCMI CPUFreq. This is a pure ARM stuff. Do we need to keep it in a common code and build it for other than ARM platforms?

I see that corresponding headers are for ARM only:  
xen/include/asm-arm/kernel.h
xen/include/asm-arm/vscmi.h

lorc replied:
Yes, this is a good idea.

otyshchenko1 commented on xen/drivers/cpufreq/cpufreq_vscmi.c:
```diff
@@ -0,0 +1,115 @@
+#include <xen/init.h>
+#include <xen/percpu.h>
+#include <xen/sched.h>
+#include <xen/cpufreq.h>
```
Please put all files in alphabetic order.

otyshchenko1 commented on xen/drivers/cpufreq/cpufreq_vscmi.c:
```diff
@@ -0,0 +1,115 @@
+                if ( requsted_opp < v->arch.opp )
+                    requsted_opp = v->arch.opp;
+            }
+        }
```
Can we optimize such 3 level loop somehow?

lorc replied:
Any ideas how?

otyshchenko1 commented on xen/arch/arm/vscmi.c:
```diff
@@ -0,0 +1,356 @@
+
+int domain_vscmi_init(struct domain *d, gfn_t shmem_gfn)
+{
+    d->arch.scmi_base_pg = alloc_domheap_page(d, 0);
```
alloc_domheap_page may fail...

otyshchenko1 commented on xen/arch/arm/vscmi.c:
```diff
@@ -0,0 +1,356 @@
+           d->arch.scmi_base_ipa,
+           page_to_maddr(d->arch.scmi_base_pg));
+
+    map_regions_p2mt(d, shmem_gfn, 1,
```
map_regions_p2mt may fail as well...

otyshchenko1 commented on xen/arch/arm/domain_build.c:
```diff
@@ -1929,6 +2112,9 @@ static void __init find_gnttab_region(struct domain *d,
     printk("Grant table range: %#"PRIpaddr"-%#"PRIpaddr"\n",
            kinfo->gnttab_start, kinfo->gnttab_start + kinfo->gnttab_size);
+
+    /* TODO: Move this to own function */
```
Who and when will address this?

lorc replied:
We. In some future. Do you prefer to not leave souch `TODO` comments?

otyshchenko1 replied:
No, it's ok.

otyshchenko1 commented on xen/arch/arm/vscmi.c:
```diff
@@ -0,0 +1,356 @@
+    unmap_regions_p2mt(d, gaddr_to_gfn(d->arch.scmi_base_ipa), 1,
+                       page_to_mfn(d->arch.scmi_base_pg));
+
+    free_domheap_page(d->arch.scmi_base_pg);
```
Do we need to set to NULL here?
d->arch.scmi_base_pg = NULL;

lorc replied:
No, because this function is called in destruction chain of domain. I can't see how it can be used after free.

otyshchenko1 replied:
got it

otyshchenko1 commented on xen/arch/arm/Makefile:
```diff
@@ -60,6 +60,7 @@ obj-y += vtimer.o
 obj-$(CONFIG_SBSA_VUART_CONSOLE) += vpl011.o
 obj-y += vsmc.o
 obj-y += vpsci.o
+obj-y += vscmi.o
```
Do we need to put VSCMI stuff under CONFIG_VSCMI? I think, there should be a build option to let user to have this feature configurable.

lorc replied:
Yeah, I think so. Will do it.

otyshchenko1 commented on xen/arch/arm/vscmi.c:
```diff
@@ -0,0 +1,356 @@
+
+int vcpu_vscmi_init(struct vcpu *vcpu)
+{
+    vcpu->arch.opp = 0;
```
Please consider adding some define for this. It is not completely clear whether 0 is default or undefined/error value here.

lorc replied:
It is still under development. Do you checked commit messages? I described this change there.

otyshchenko1 commented on xen/arch/arm/vscmi.c:
```diff
@@ -0,0 +1,356 @@
+{
+    struct scmi_shared_mem *data;
+    uint32_t hdr;
+    int res;
```
I would use local variable scmi_base_pg = current->domain->arch.scmi_base_pg; to not write long line many times down the function. The same could be applied for scmi_base_ipa also

otyshchenko1 commented on xen/arch/arm/domain_build.c:
```diff
@@ -2161,6 +2347,8 @@ int __init construct_dom0(struct domain *d)
     if ( rc < 0 )
         return rc;
 
+    rc = domain_vscmi_init(d, gaddr_to_gfn(kinfo.scmi_shmem));
```
Please check the return value here. There will be a few reasons for domain_vscmi_init to fail

lorc replied:
Good catch, thanks.

otyshchenko1 commented on xen/arch/arm/vscmi.c:
```diff
@@ -0,0 +1,356 @@
+int domain_vscmi_init(struct domain *d, gfn_t shmem_gfn)
+{
+    d->arch.scmi_base_pg = alloc_domheap_page(d, 0);
+    d->arch.scmi_base_ipa = gfn_to_gaddr(shmem_gfn);
```
I would check shmem_gfn if possible before taking into the accounting. At least to be sure it is a valid value and really belongs to this domain. I wouldn't 100% trust shmem_gfn supplied by libxl for example.

lorc replied:
Whoa, how `gfn` can belong to some other guest? It is valid only in context of a certain domain.
And against exactly what you want me to check it?

otyshchenko1 replied:
I would check against INVALID_GFN then:
if ( gfn_eq(gfn, INVALID_GFN) )
...

I will leave it to you to decide.

otyshchenko1 commented on xen/arch/arm/vscmi.c:
```diff
@@ -0,0 +1,356 @@
+/*
```
A lot of Xen coding style errors across this file.

otyshchenko1 commented on xen/arch/arm/vscmi.c:
```diff
@@ -0,0 +1,356 @@
+    case PERF_LEVEL_SET:
+    {
+        struct scmi_perf_set_level *req = (void*)data->msg_payload;
+        int domain = le32_to_cpu(req->domain);
```
Is it a Xen domain or SCMI power/voltage domain? 

I am asking because the following line of code looks a bit confusing:
current->domain->vcpu[domain]->arch.opp = level;

Please consider naming it appropriate to avoid confusing...

lorc replied:
this is SCMI perf domain, as in SCMI protocol.

otyshchenko1 commented on xen/arch/arm/vscmi.c:
```diff
@@ -0,0 +1,356 @@
+        return false;
+    }
+
+    data = (struct scmi_shared_mem*)xzalloc_array(char, 256);
```
What does '256' mean? I see this magic being used at least three times across this file.

lorc replied:
Yes, good catch. We need to settle with some defined shared buffer size.

otyshchenko1 replied:
Also, I think we can drop casting here (struct scmi_shared_mem*)

otyshchenko1 commented on xen/arch/arm/vscmi.c:
```diff
@@ -0,0 +1,356 @@
+    flush_page_to_ram(page_to_mfn(current->domain->arch.scmi_base_pg), false);
+    if ( res )
+    {
+        printk(XENLOG_ERR "Error writing guest memory %d\n", res);
```
gdprintk?

lorc replied:
`gprintk`

otyshchenko1 commented on xen/arch/arm/vscmi.c:
```diff
@@ -0,0 +1,356 @@
+                                     data, 256, false);
+    if ( res )
+    {
+        printk(XENLOG_ERR "Error reading guest memory %d\n", res);
```
gdprintk?

lorc replied:
`gprintk`, I believe

otyshchenko1 commented on xen/arch/arm/vscmi.c:
```diff
@@ -0,0 +1,356 @@
+        goto err;
+    }
+
+    xfree(data);
```
You free data whether error or not. Maybe something like:

err:
    xfree(data);
    return !res;

lorc replied:
Good idea, thanks

otyshchenko1 commented on xen/arch/arm/vscmi.c:
```diff
@@ -0,0 +1,356 @@
+        }
+        write_32(data->msg_payload, SCMI_SUCCESS);
+
+        attrs->flags = cpu_to_le32(BIT(31) | BIT(30));
```
Common to handle_base_req and handle_perf_req:
Please  minimize the usage of different magics as possible.

otyshchenko1 commented on xen/drivers/cpufreq/cpufreq_vscmi.c:
```diff
@@ -0,0 +1,115 @@
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
```
I am afraid, something is missed here. Please follow the document http://xenbits.xen.org/gitweb/?p=xen.git;a=blob;f=CONTRIBUTING

lorc replied:
Oh, okay

otyshchenko1 commented on xen/drivers/cpufreq/cpufreq_vscmi.c:
```diff
@@ -0,0 +1,115 @@
+{
+    struct vcpu *vcpu = hcpu;
+    struct cpufreq_policy *policy;
+    int requsted_opp = 0;
```
s/requsted_opp/requested_opp

otyshchenko1 commented on xen/include/asm-arm/vscmi.h:
```diff
@@ -0,0 +1,48 @@
+}
+
+#endif
+
```
/*
 * Local variables:
 * mode: C
 * c-file-style: "BSD"
 * c-basic-offset: 4
 * indent-tabs-mode: nil
 * End:
 */

otyshchenko1 commented on xen/arch/arm/vscmi.c:
```diff
@@ -0,0 +1,356 @@
+            data->length = sizeof(uint32_t) * 2;
+            break;
+        }
+
```
Do we need to take an action if we are already at the required perf level?

if (current->domain->vcpu[domain]->arch.opp == level)

lorc committed:
xen/pm: add support for guest_pm module

guest_pm provides per-guest power management policies.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
xen/vscmi: move number of OPPs to guest_pm code

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
vscmi: force enabled guest_pm

If users wishes to enable vSCMI server, this means that they
want to allow guests to control PM state of a platform.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
cpufreq_governor_vscmi: add guest_pm policies support

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
tools: add support for guest_pm runtime control

add "xl guest-pm-show" and "xl guest-pm-set" commands

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
tools: add guestpm option to xl.cfg

This option enables guestpm configuration and domain creation
time.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
xen/arm: add command line option to disable vscmi for dom0

New command-line option 'dom_vscmi' controls whenever Dom0
can issue SCMI calls. It is enabled by default.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
xen/arm: cpufreq_vscmi: handle possible error in cpufreq

Handle return value of __cpufreq_driver_target(). As we can't
do much, we will just report the error code and return from
callback handler.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
xen/arm: vscmi: add unregister_vscmi_notifier() function

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
xen/arm: guest_pm: add unregister_gust_pm_notifier() function

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
xen/arm: cpufreq_vscmi: handle start/stop actions

Now governor will handle register notifier callbacks
on start and un-regiter them on stop.

As start action can be called multiple times (one time
per pCPU) we need some counter to make sure that
we register callbacks only once.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
xen/arm: vscmi: handle error during calling notifiers

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
xen/arm: domain_build: move scmi shmem initialization

find_gnttab_region() is not the best place to initialize
vSCMI memory region. Better to do this in construct_dom0().

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
xen/arm: vscmi: introduce VSCMI_SHM_SIZE

To remove magic constants, add definition for vSCMI shared buffer
size.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc commented:
@otyshchenko1, could you please merge this PR?

lorc commented:
@iartemenko, could you please merge?
